File: apps/api/gentletap/middleware/security_headers.py
"""Add baseline security headers to every API response."""

from collections.abc import Awaitable, Callable

from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response

SECURITY_HEADERS: dict[str, str] = {
    "X-Content-Type-Options": "nosniff",
    "X-Frame-Options": "DENY",
    "Referrer-Policy": "strict-origin-when-cross-origin",
    "Permissions-Policy": "camera=(), microphone=(), geolocation=(), payment=(), usb=()",
    "X-XSS-Protection": "1; mode=block",
    "Cache-Control": "no-store",
}


class SecurityHeadersMiddleware(BaseHTTPMiddleware):
    async def dispatch(
        self,
        request: Request,
        call_next: Callable[[Request], Awaitable[Response]],
    ) -> Response:
        response = await call_next(request)
        for key, value in SECURITY_HEADERS.items():
            if key not in response.headers:
                response.headers[key] = value
        return response

File: apps/web/src/lib/security-headers.ts
/** Shared HTTP security headers for Next.js and documentation. */

const isProd = process.env.NODE_ENV === "production";

/**
 * CSP tuned for Next.js (inline hydration scripts) + same-origin API proxy.
 * Paddle.js overlay checkout loads scripts, an iframe, and XHRs from *.paddle.com
 * (and *.paddleedge.com for the sandbox/profiler). The hosted-URL fallback uses
 * top-level navigation and needs no CSP entry.
 */
const PADDLE_HOSTS = "https://*.paddle.com https://*.paddleedge.com";

export const CONTENT_SECURITY_POLICY = [
  "default-src 'self'",
  "base-uri 'self'",
  "object-src 'none'",
  "frame-ancestors 'none'",
  "form-action 'self'",
  `script-src 'self' 'unsafe-inline' 'unsafe-eval' ${PADDLE_HOSTS}`,
  `style-src 'self' 'unsafe-inline' ${PADDLE_HOSTS}`,
  `img-src 'self' data: blob: ${PADDLE_HOSTS}`,
  "font-src 'self' data:",
  `connect-src 'self' ${PADDLE_HOSTS}`,
  `frame-src ${PADDLE_HOSTS}`,
  ...(isProd ? ["upgrade-insecure-requests"] : []),
].join("; ");

export const SECURITY_HEADER_ENTRIES: { key: string; value: string }[] = [
  { key: "Content-Security-Policy", value: CONTENT_SECURITY_POLICY },
  { key: "X-Frame-Options", value: "DENY" },
  { key: "X-Content-Type-Options", value: "nosniff" },
  { key: "Referrer-Policy", value: "strict-origin-when-cross-origin" },
  {
    key: "Permissions-Policy",
    value: "camera=(), microphone=(), geolocation=(), payment=(), usb=()",
  },
  // Legacy browsers; modern guidance prefers CSP. Kept for scanner compliance.
  { key: "X-XSS-Protection", value: "1; mode=block" },
];
